### PR TITLE
Align Helm/Ansible plugins using common/v3 base 

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**
- Align Helm/Ansible with Go by making use of the common base and by removing the duplications
- Align SDK with KB
- Add configComponent feature for Ansible/Helm
- Add a new plugin option to let users create only the common base with sdk

**Motvation**
Closes: https://github.com/operator-framework/operator-sdk/issues/4542
Closes: https://github.com/operator-framework/operator-sdk/issues/4643
 
**Blocked by**
https://github.com/kubernetes-sigs/kubebuilder/pull/2106

### CODE REVIEW

File: testdata/go/v2/memcached-operator/Makefile
Comment (camilamacedo86) [inline testdata/go/v2/memcached-operator/Makefile:86]: The issue is only in the master branch.
This PR fixes it.

File: hack/generate/samples/internal/ansible/memcached.go
Comment (Adirio) [inline hack/generate/samples/internal/ansible/memcached.go:22]: Import order

File: hack/generate/samples/internal/ansible/memcached_molecule.go
Comment (Adirio) [inline hack/generate/samples/internal/ansible/memcached_molecule.go:23]: Import order

File: internal/cmd/operator-sdk/cli/cli.go
Comment (Adirio) [inline internal/cmd/operator-sdk/cli/cli.go:104]: I would not include this for now as it is not used for anything and it will imply backwards compatibility issues if removed later.

File: internal/plugins/ansible/v1/init.go
Comment (Adirio) [inline internal/plugins/ansible/v1/init.go:22]: Import order

File: internal/plugins/ansible/v1/init.go
Comment (Adirio) [inline internal/plugins/ansible/v1/init.go:120]: This should not be done in post-scaffold, this should be done in scaffold hook.

File: internal/plugins/helm/v1/init.go
Comment (Adirio) [inline internal/plugins/helm/v1/init.go:143]: Same, in scaffold and not post-scaffold

File: internal/testutils/olm.go
Comment (Adirio) [inline internal/testutils/olm.go:22]: Import order

File: internal/util/utils.go
Comment (Adirio) [inline internal/util/utils.go:26]: As we are going to use this functions to scaffold, they should receive a `filesystem.Filesystem` and use `afero` internally instead of `ioutil` or `os`. If we use these others directly, once we implement improvements in the filesystem abstraction (for example the scaffold atomicity we talked about) these methods will fail.

File: internal/util/utils.go
Comment (Adirio) [inline internal/util/utils.go:46]: Same

File: internal/util/utils.go
Comment (Adirio) [inline internal/util/utils.go:71]: Same

File: test/integration/suite_test.go
Comment (Adirio) [inline test/integration/suite_test.go:24]: Import order

**Review by Adirio — CHANGES_REQUESTED** [review]

File: internal/util/utils.go
Comment (camilamacedo86) [reply internal/util/utils.go:26]: We can do that in a follow-up. This funcs or part of them exist in KB https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/utils/util.go#L63

Note that there or here we might Insert/Replace data in the files. So, IMO it can be promoted as utils from KB. 
Then, we can check the possibility to change that to use afero.

File: internal/util/utils.go
Comment (camilamacedo86) [reply internal/util/utils.go:46]: see: https://github.com/operator-framework/operator-sdk/pull/4701#discussion_r602166043
I hope that is fine we close this one since is duplicated.

File: internal/util/utils.go
Comment (camilamacedo86) [reply internal/util/utils.go:71]: see: https://github.com/operator-framework/operator-sdk/pull/4701#discussion_r602166043
I hope that is fine we close this one since is duplicated.

File: internal/plugins/envtest/v1/init.go
Comment (camilamacedo86) [inline internal/plugins/envtest/v1/init.go:58]: @Adirio , regarding the code implementation, suggested which I am ok within a follow-up. See that it was NOT raising the issue when the target is not found as it is in the Insert/replace functions that we used. Just to let, you know.

Comment (camilamacedo86) [reply internal/plugins/envtest/v1/init.go:58]: It needs to be done in another pr.

File: internal/cmd/operator-sdk/cli/cli.go
Comment (camilamacedo86) [reply internal/cmd/operator-sdk/cli/cli.go:104]: It is fine. It is only to show @jmrodri the possibility of something that we spoke about. That would be in its own pr with my wish/idea of adding a Plugin Contribution doc with the vision of plugins. Also, we need/can add the custom plugin provide by kB ( the declarative.go.kubebuilder.io/v1: pattern too to allow people to get advantage of this one and custom their implementation )

It allows people such as Java developers who want to create their own plugin to follow up the same base structure and then scaffold the rest with their language of preference instead of using Golang. 

c/c

File: internal/util/utils.go
Comment (camilamacedo86) [reply internal/util/utils.go:26]: Also see that we have a very complex code in the manifests plugin that could be simplified with that.

Comment (Adirio) [reply internal/util/utils.go:26]: But before this PR, and in the KB package you mentioned, they are used during tests after running the correspondig command call, which means that the file exists in disk.

However, here you are using these functions during scaffolding, and all scaffolding should be done through the filesystem abstraction, not directly with `os` and `ioutil` packages. Thats the point I wanted to make. I know the current machinery only allows to create fields from templates, or insert code-fragments before special markers. Do we need to improve the machinery for generalized insertion and replacement? Yes. But until that is done, SDK will have to rely on these functions, and scaffolding directly to the filesystem without using the abstraction provided should not be done.

Comment (camilamacedo86) [reply internal/util/utils.go:26]: It is not about the file. It is about the content not be found. The same issue can happen there indeed in the tests. So, for example,  the test will pass because the required change has not been performed. 

However, I am totally fine with the improvements sugggested getting done in follow ups. the intention here was just to share the scenario.

Comment (Adirio) [reply internal/util/utils.go:26]: I'm not talking about the logic if the content is not found Camila, I'm talking about the filesystenm abstraction. We are providing one common filesystem abstraction that all `Scaffold()` **must** use. This allows us to, for example, test by saying that the underlying filesystem is an `afero.MemMapFs`. If we do this, when it gets to this scaffold, the part that uses the system abstraction will work, but the direct calls to `os` and `ioutil` will break. If you want to use these functions inside `Scaffold` and `PreScaffold` methods, you must turn those calls to `os` and `ioutil` functions to the equivalents of `afero` and pass the filesystem abstraction. If you only want to use them for tests (like until now) you don't have this problem. After moving everything to `afero`, when you call this function from tests that don't have a filesystem you just create one like so: `filesystem.Filesystem{FS: afero.NewOsFs()}`.

Comment (camilamacedo86) [reply internal/util/utils.go:26]: Could not it be done in a follow-up? Could we not done this after we improve these implementations upstream and provide them from there? Is that a blocker for we to move forward here?

Comment (Adirio) [reply internal/util/utils.go:26]: I fear it should block, but should be pretty easy:
- Add `fs filesystem.Filesystem` as an argument.
- When calling this function from inside `Scaffold` pass the `fs` that you have there.
- When calling this function from any other place (tests) pass it a `filesystem.Filesystem(FS: afero.NewOsFs())`.
- When using `ioutil.Xxxxxx(...)` change it to `afero.Xxxxxx(fs.FS, ...)`.
- When using `os.Xxxxxx(....)` change it to `fs.FS.Xxxxxx(...)`.

File: changelog/fragments/add_common_kb.yaml
Comment (camilamacedo86) [inline changelog/fragments/add_common_kb.yaml:46]: it will be removed

File: internal/cmd/operator-sdk/cli/cli.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/cli/cli.go:85]: We might not need to use the common here. see: https://github.com/kubernetes-sigs/kubebuilder/pull/2112

Comment (Adirio) [reply internal/cmd/operator-sdk/cli/cli.go:85]: In order to do this, we need to export the bundle, which I have planned to do once kubernetes-sigs/kubebuilder#2106 is merged. (I'm not doing it right now to prevent rebases)

Comment (camilamacedo86) [reply internal/cmd/operator-sdk/cli/cli.go:85]: we can do that in a follow-up. I will add a comment

File: internal/util/utils.go
Comment (camilamacedo86) [reply internal/util/utils.go:26]: What bug/issue we will face that justify we block this PR instead of address that in follow up when your pr https://github.com/kubernetes-sigs/kubebuilder/pull/2119 get merged?

**Review by jmrodri — APPROVED** [review]
/lgtm

File: changelog/fragments/add_common_kb.yaml
Comment (camilamacedo86) [inline changelog/fragments/add_common_kb.yaml:5]: My only concern here is; Have any reason for we do not add this option for Ansible/Helm as well? 
c/c @jmrodri @estroz

File: internal/plugins/ansible/v1/scaffolds/internal/templates/config/kdefault/kustomization.go
Comment (estroz) [inline internal/plugins/ansible/v1/scaffolds/internal/templates/config/kdefault/kustomization.go:42]: I think this change can be reverted.

File: internal/plugins/helm/v1/scaffolds/internal/templates/config/kdefault/kustomization.go
Comment (estroz) [inline internal/plugins/helm/v1/scaffolds/internal/templates/config/kdefault/kustomization.go:42]: Ditto

File: internal/plugins/ansible/v1/init.go
Comment (estroz) [inline internal/plugins/ansible/v1/init.go:193]: Small nit to align YAML formatting
```suggestion
        env:
        - name: ANSIBLE_GATHERING
          value: explicit`
```

File: internal/plugins/ansible/v1/scaffolds/internal/templates/config/kdefault/kustomization.go
Comment (estroz) [reply internal/plugins/ansible/v1/scaffolds/internal/templates/config/kdefault/kustomization.go:42]: Ah nevermind, the base kustomization.yaml is configured with stuff an ansible operator (and helm operator) can't really use out of the box.

**Review by estroz — APPROVED** [review]
LGTM other than the formatting nit

/lgtm

File: internal/plugins/ansible/v1/init.go
Comment (camilamacedo86) [reply internal/plugins/ansible/v1/init.go]: done

**Pull request merged**
